backend/src/main/java/com/javahelp/backend/domain/user/register/UserRegisterInteractor.java

```diff
@@ -0,0 +1,126 @@
+package com.javahelp.backend.domain.user.register;
+
+import com.javahelp.backend.data.IUserStore;
+import com.javahelp.model.user.User;
+import com.javahelp.model.user.UserInfo;
+import com.javahelp.model.user.UserPassword;
+
+/**
+ * An interactor to make new {@link User} and {@link UserPassword} for all types of users
+ */
+public abstract class UserRegisterInteractor <T extends IUserRegisterInputBoundary,
+        U extends UserInfo>{
+
+    IUserStore userStore;
+
+    /**
+     * Creates a new {@link UserRegisterInteractor} instance
+     *
+     * @param userStore  {@link IUserStore} to use
+     */
+    public UserRegisterInteractor(IUserStore userStore) {
+        this.userStore = userStore;
+    }
+
+    /**
+     * Generate a new pseudo {@link User}. Private because this is exclusively for a general
+     * {@link User} type.
+     *
+     * @return a new {@link User} entity
+     */
+    private User register(T boundary){
+        return new User("", this.bundleInfo(boundary), boundary.getUsername());
+    }
+
+    /**
+     * Checking for violations for registering a {@link User} account.
+     * @param boundary an instance of {@link IUserRegisterInputBoundary} to retrieve input info
+     *
+     * @return a {@link RegisterUserResponse} model that contains message of failure at registration
+     * Null if all tests passed.
+     */
+    protected RegisterUserResponse isValidRegistering(T boundary){
```

Comment:
This method should have a more specific name, if it checks for email and username it should be called something like `emailAndUsernameTaken`, etc...

Reply:
I also don't see a specific need for this to be protected rather than private

Reply:
You're exactly right